### PR TITLE
[REEF-498] Create scripts for release versioning, packaging, and checksum

### PR DESCRIPTION
**Pull request by dongjun-Lee**
This pull request addressed the issue by
* creating `release.py` which creates a release candidate and relevant files.
* creating `change_version.py` which changes versions in every pom.xml and relevant files.

JIRA: [REEF-498](https://issues.apache.org/jira/browse/REEF-498)

### CODE REVIEW

File: dev/change_version.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    f = open(file, 'r')
+
+    while True:
+        line = f.readline()
+        if "<version>" in line:
+            break
+        changed_str += line
+
+    r = re.compile('<version>(.*?)</version>')
```
Comment (markusweimer) [inline dev/change_version.py]: Is this safe? Worst case, this picks up additional lines in the pom. Can we parse the XML instead?

File: dev/change_version.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            changed_str += line.replace(old_version, new_version)
+        else:
+            changed_str += line
+
+    f = open(file, 'w')
+    f.write(changed_str)
+    f.close()
+
+"""
+Change the name of shaded.jar in lang/cs/Org.Apache.REEF.Client/run.cmd
```
Comment (markusweimer) [inline dev/change_version.py]: I believe `run.cmd` is deprecated. @jwang98052, correct?

File: dev/change_version.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (markusweimer) [inline dev/change_version.py]: This script also needs to change the versions found in `AssemblyInfo.cs` as well as the NuGet versions.

File: dev/change_version.py (hunk 1 above)
Comment (dongjun-Lee) [inline dev/change_version.py]: @markusweimer Thank you for your comment. At first, I parsed XML and rebuilt it but it ignored comments and made unnecessary changes such as indent or newline. I think this is quite safe because change happens only once for a file.

**Comment by dongjun-Lee** [comment]
I changed the script to change build.props and every AssemblyInfo.cs files. Since parsing XMLs and rebuilding have problems like ones I've mentioned above, instead of changing the parsing process I've changed the existing method to be safer.
@markusweimer @dafrista , could you please take a look at this commit? Thank you.

File: dev/change_version.py (hunk 2 above)
Comment (jwang98052) [inline dev/change_version.py]: It is still used when running REEF.Tests on Yarn.

File: dev/release.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        gpg_result = subprocess.call(gpg_str, shell=True)
+
+        if gpg_result == 0:
+            md5 = hashlib.md5(open(file_name,'rb').read()).hexdigest()
+            sha = hashlib.sha512(open(file_name,'rb').read()).hexdigest()
+
+            md5_file = open(file_name + ".md5", "w")
+            md5_file.write("MD5(" + file_name + ")= " + md5 + "\n")
+            sha_file = open(file_name + ".sha", "w")
+            sha_file.write("SHA(" + file_name + ")= " + sha + "\n")
```
Comment (dafrista) [inline dev/release.py]: It looks like the output is meant to follow `openssl`. But it's not quite the same. Replacing `SHA(` with `SHA512(` will do that.

(We're still discussing at dev@ whether following `openssl` should be the correct output format.)

Comment (dafrista) [inline dev/release.py]: If there are no objections, let's change to "coreutils style". This is the style [currently used on rc2](https://dist.apache.org/repos/dist/dev/incubator/reef/0.12.0-incubating-rc2/), the checksum followed by a space, asterisk, then the filename. (The asterisk means that the file being checksum-ed was treated as binary, which is what we want.)

Comment (markusweimer) [inline dev/release.py]: +1 on coreutils. They are widely available and we can make PowerShell equivalents

**Comment by dongjun-Lee** [comment]
@dafrista Thank you for the comment. I addressed it.

File: dev/release.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    return_str += "Checksums of apache-reef-" + reef_version + "-rc" + str(rc_num) + ".tar.gz:\n\n"
+
+    md5 = open(file_name + ".md5").read().split(" ")[0]
+    return_str += "MD5: " + md5 + "\n"
+    
+    sha = open(file_name + ".sha512.txt").read().split(" ")[0]
+    return_str += "SHA: " + sha + "\n"
+
+    return_str += "\nRelease artifacts are signed with the key. The KEYS file is available here:\n"
```
Comment (markusweimer) [inline dev/release.py]: `the key` sounds odd here. Maybe a better formulation would be:

```
Release artifacts are signed with a key found in the KEYS file available here:
```

**Comment by dongjun-Lee** [comment]
Thanks @markusweimer, I fixed it.

File: dev/change_version.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+This script changes versions in every pom.xml and relevant files.
+
+(How to run)
+python change_version <reef_home> <reef_version_for_pom.xml>  (optional) -s <true or false>   
+
+-s option changes value of 'IsSnapshot' in lang/cs/build.props.
+If you use the option "-s false", bulid.props changes as,
+ <RemoveIncubating>true</RemoveIncubating>
+ <IsSnapshot>false</IsSnapshot>
+ <SnapshotNumber>0</SnapshotNumber>
```
Comment (dafrista) [inline dev/change_version.py]: @markusweimer the snapshot uses the `##` format (two digits), right? If so, we should change this to show e.g. 00.

File: dev/change_version.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            break
+        if "<RemoveIncubating>" and "</RemoveIncubating>" in line and is_snapshot=="false":
+            old_remove_incubating = r1.search(line).group(1)
+            changed_str += line.replace(old_remove_incubating, "true")
+        elif "<IsSnapshot>" in line and "</IsSnapshot>" in line:
+            old_is_snapshot = r2.search(line).group(1)
+            changed_str += line.replace(old_is_snapshot, is_snapshot)
+        elif "<SnapshotNumber>" in line and "</SnapshotNumber>" in line and is_snapshot=="false":
+            old_snapshot_number = r3.search(line).group(1)
+            changed_str += line.replace(old_snapshot_number, "0")
```
Comment (dafrista) [inline dev/change_version.py]: See above comment. This might have to be `"00"`.

File: dev/change_version.py (hunk 6 above)
Comment (markusweimer) [inline dev/change_version.py]: I believe so. We need to generate a unique version each time we want to publish a NuGet SNAPSHOT. 99 snapshots between releases seems more realistic than 9 :-)

File: dev/change_version.py (hunk 7 above)
Comment (jwang98052) [inline dev/change_version.py]: Yes. Unique and allow up to 99 was the original reason.

**Comment by dongjun-Lee** [comment]
@dafrista I fixed it. Thank you for the comment.

**Comment by markusweimer** [comment]
Given that we've done the release, is it time to merge this?

**Comment by dafrista** [comment]
Thanks for asking!

This has been a classic example of changing specifications... and I'd like to ask for one more. When changing version numbers for the maven release, *only* the poms must be changed. I accomplished this by commenting out stuff in the script. But, it would make sense to add such an option. @dongjun-Lee is this possible? Thanks for bearing with me.

Once that gets in, we can merge. Then I'll change the cwiki entries to add how some of the steps can be made easy via the scripts.

**Comment by dongjun-Lee** [comment]
Thank you @dafrista . I added an option to change only the poms.

**Comment by dafrista** [comment]
@dongjun-Lee I've updated the [release manager guide](https://cwiki.apache.org/confluence/display/REEF/Release+Manager+Guide). Please double-check the script instructions that I've added.

**Comment by dongjun-Lee** [comment]
@dafrista Sorry, I made a mistake. I fixed the bug.
Thanks a lot for the guide. In the guide, Step 3 might be included in step 4 because `release.py` includes removing "-SNAPSHOT" and changing versions in all files.

**Comment by dafrista** [comment]
@dongjun-Lee thanks for the update! It looks good.

Can we remove the use of `change_version` in `release`?

I'm probably contradicting an earlier request. However, I now think this amount of automation is dangerous. The release manager should tarball the exact changes he made on the branch.

After this final change, I'm happy to merge.

File: dev/release.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+python release.py ~/incubator-reef 0.12.0-incubating 1 E488F925
+
+"""
+
+
+import subprocess
+import fnmatch
+import tarfile
+import hashlib
+import change_version
```
Comment (dafrista) [inline dev/release.py]: Let's remove this (more details in the previous comment).

**Comment by dongjun-Lee** [comment]
@dafrista I removed it. Thank you :)
